fdbserver/workloads/WatchesSameKeyCorrectness.actor.cpp

```diff
@@ -0,0 +1,232 @@
+/*
+ * WatchesSameKeyCorrectness.actor.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbrpc/ContinuousSample.h"
+#include "fdbclient/ReadYourWrites.h"
+#include "fdbclient/NativeAPI.actor.h"
+#include "fdbserver/TesterInterface.actor.h"
+#include "flow/DeterministicRandom.h"
+#include "fdbserver/workloads/workloads.actor.h"
+#include "flow/actorcompiler.h"  // This must be the last #include.
+
+struct WatchesSameKeyWorkload : TestWorkload {
+	int numWatches;
+	std::vector<Future<Void>> cases;
+
+	WatchesSameKeyWorkload(WorkloadContext const& wcx)
+		: TestWorkload(wcx)
+	{
+		numWatches = getOption( options, LiteralStringRef("numWatches"), 3 );
+	}
+
+	std::string description() const override { return "WatchesSameKeyCorrectness"; }
+
+	Future<Void> setup(Database const& cx) override { 
+		if ( clientId == 0 ) {
+			cases.push_back( case1(cx, LiteralStringRef("foo1"), this) );
+			cases.push_back( case2(cx, LiteralStringRef("foo2"), this) );
+			cases.push_back( case3(cx, LiteralStringRef("foo3"), this) );
+			cases.push_back( case4(cx, LiteralStringRef("foo4"), this) );
+			cases.push_back( case5(cx, LiteralStringRef("foo5"), this) );
+		}
+		return Void();
+	}
+
+	Future<Void> start(Database const& cx) override { 
+		if (clientId == 0) return waitForAll( cases );
+		return Void();
+	}
+
+	Future<bool> check(Database const& cx) override {
+		if ( clientId != 0 ) return true;
+		bool ok = true;
+		for( int i = 0; i < cases.size(); i++ ) {
+			if ( cases[i].isError() ) ok = false;
+		}
+		cases.clear();
+		return ok;
+	}
+
+	ACTOR static Future<Void> setKeyRandomValue(Database cx, Key key, Optional<Value> val) {
+		try {
+			if ( !val.present() ) val = Value( deterministicRandom()->randomUniqueID().toString() );
+			state ReadYourWritesTransaction tr(cx);
+			tr.set(key, val.get());
+			wait( tr.commit() );
+			return Void();
+		} catch ( Error& e ) {
+			throw e;
+		}
+	}
+
+	ACTOR static Future<Optional<Value>> getValue(Database cx, Key key) {
+		try {
+			state ReadYourWritesTransaction tr(cx);
+			Optional<Value> val = wait(tr.get(key));
+			return val;
+		} catch ( Error& e ) {
+			throw e;
+		}
+	}
+
+	ACTOR static Future<Future<Void>> watchKey(Database cx, Key key) {
+		try {
```

Comment:
Please use the standard retry loop here. When you run a transaction in FDB you almost always should write it like this:

```
state ReadYourWritesTransaction tr(cx);
loop {
    try {
        // the transaction logic comes here
       return Void(); // will only be reached if the transaction succeeds
    } catch (Error& e) {
        wait(tr.onError(e));
    }
}
```